### DISCUSSION TITLE
Code suggestion and imports not properly working

### DISCUSSION THREAD

**Discussion by PUNK3DAF** [apache/netbeans#6963] [Q&A]
Hello NetBeans Community,

I hope this message finds you well. I am currently facing a unique issue with my NetBeans IDE and JDK installation, and I'm seeking assistance from the community.

Problem:
I have recently installed JDK 17.0.1, and I am experiencing difficulties with NetBeans recognizing the javax.swing.table and java.sql packages. Upon investigating the JDK installation, I noticed an unusual file named modules in the lib directory instead of the expected modules folder (at least I've been told it should be a folder, not a file). When I create a class that has a Superclass AbstractTableModel, I can't generate a constructor, and when I use methods from java.sql classes, the import suggestions don't suggest java.sql classes.

Steps Taken:

Reinstalled JDK 17.0.1 with the complete installation package.
Verified the presence of the modules file in the lib directory, which appears to be a binary file.
Set the JAVA_HOME environment variable to the correct JDK path.
Updated NetBeans configuration to use JDK 17.0.1 in "Tools" -> "Java Platforms."
Performed a clean and build operation on the NetBeans project.
Restarted NetBeans after making changes.
Despite these efforts, the issue persists, and NetBeans is not recognizing the javax.swing.table and java.sql packages.

System Information:

Operating System: Windows 11 x64
NetBeans Version: 14
JDK Version: 17.0.1

Additional Notes:
I have already sought assistance from various forums and reinstalled NetBeans multiple times without success.
The community is kindly requested to provide insights, suggestions, or solutions to resolve this issue.
Any help or guidance you can provide would be greatly appreciated. Thank you for your time and assistance.

Best regards,
PUNK3DAF

**Comment by matthiasblaesing** [comment]
Please provide a minimal sample of a project, that reproduces the problem.

**Reply by PUNK3DAF** [reply]
Hi matthiasblaesing,

Thank you for your prompt response and willingness to help. Below, you'll find the relevant code snippets:

```java
package base;

public class Connect {

    private Connection connection; 
    // The following line should ideally trigger an import suggestion for java.sql.Connection,
    // but it's not happening automatically.

}
```
```java
package test;

import javax.swing.table.AbstractTableModel;

public class TestAbstractClass extends AbstractTableModel {

    // The IDE should typically provide an option to generate constructors automatically,
    // but it's not happening in this case.

    @Override
    public int getRowCount() {
        throw new UnsupportedOperationException("Not supported yet."); // Generated from nbfs://nbhost/SystemFileSystem/Templates/Classes/Code/GeneratedMethodBody
    }

    @Override
    public int getColumnCount() {
        throw new UnsupportedOperationException("Not supported yet."); // Generated from nbfs://nbhost/SystemFileSystem/Templates/Classes/Code/GeneratedMethodBody
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex) {
        throw new UnsupportedOperationException("Not supported yet."); // Generated from nbfs://nbhost/SystemFileSystem/Templates/Classes/Code/GeneratedMethodBody
    }
}
```

**Comment by matthiasblaesing** [comment]
This sample can't reproduce your problem:

![image](https://github.com/apache/netbeans/assets/2179736/76fc6c4c-a5db-4a8d-8837-a45539639d90)

It is important to create a **_full sample project_** and show that. In that case minimal also means, that `TestAbstractClass` should not be there as there is no apparent reason for the file to be required for reproducing your problem.

Create a minimal project, that shows your problem, use "File -> Export ZIP" and attach zip here.

**Comment by PUNK3DAF** [comment]
Thank you for your response. I've removed unnecessary files, and the project now includes only the components necessary to reproduce the problem. I've attached the ZIP file to this response.

Please let me know if there's anything else you need, and thank you for your assistance.

Best regards,
PUNK3DAF


[TestProject.zip](https://github.com/apache/netbeans/files/13970642/TestProject.zip)

**Reply by matthiasblaesing** [reply]
Thank you. The important part here is, that you used the legacy ant project type. With that information the problem is trivially reproducible:

1. Create a new Project, choose "Java with Ant" -> "Java Application"
2. Check the project properties, that a JDK 11+ is selected in Libraries and source/binary format is set to something JDK 11+
3. Insert "private Connection connection;" directly inside the body of the generated class
4. Invoke "Fix Imports"
5. Observe that this either does nothing or adds import "import com.sun.jdi.connect.spi.Connection;"
6. Create a new "Java Module Info" in the project and add `requires java.sql` to the module declaration
7. Switch back to the initially generated class and invoke "Fix Imports again"
8. Observe, that the correct import is now generated "import java.sql.Connection;"

The bug is, that the IDE assumes for JDK 11+ and ant projects, that the project is always modular, even if there is no module descriptor. It is ignored, that the application can be loaded on the classpath and that then all modules are available.

Feel free to make this an issue. As immediate help: Don't use ant projects, use Maven as basis for your project.

**Reply by PUNK3DAF** [reply]
Thank you for the detailed explanation and steps to reproduce the issue. I appreciate your efforts in helping me understand the problem.

I've confirmed that following your steps with the legacy Ant project type reproduces the issue. While I understand the recommendation to use Maven, I'd like to mention that I'm required to use Ant for this project type, as it aligns with the specifications for an upcoming university exam.

Given this constraint, I'll navigate the issue carefully and, in the meantime, adhere to the suggested workaround. I'll also explore creating an issue on the appropriate platform to report this bug, taking into account the specifics of my project requirements.

Thank you again for your assistance and guidance. If you have any further insights or recommendations, I'd greatly appreciate them.

Best regards,
PUNK3DAF
